### PR TITLE
Improve aws-lambda webhookCallback adapter

### PR DESCRIPTION
**Pull request by chrisandrewcl**
Changes:
- feat: add aws-lambda-async webhookCallback adapter
- fix: apply correct headers to aws-lambda webhookCallback adapter

I did not find any relevant test cases covering adapters, so I am opening this PR with just a few manual tests.

### CODE REVIEW

**Review by KnorpelSenf — APPROVED** [review]
Thanks! I assume that this closes #412?

Regarding the tests: it's very hard to write tests for adapters without just mocking all the APIs. But since they may change in unexpected ways anyway, there's little reason to test these few lines of code automatically. I trust your manual tests (and the community to complain if stuff breaks).

Did you test it by installing grammY from this branch using GitHub installs?

**Comment by chrisandrewcl** [comment]
I tested it on a node.js side project by switching grammy from the npm version to the fork I have on disk, which I built with `deno task backport`.

**Comment by KnorpelSenf** [comment]
> I tested it on a node.js side project by switching grammy from the npm version to the fork I have on disk, which I built with `deno task backport`.

Could you try once more using `npm install chrisandrewcl/grammY#bugfix/412-aws-lambda` just to be extra sure? I'll merge as soon as you confirm that this works. It would be awesome if you could test secret tokens, too. (See https://deno.land/x/grammy@v1.16.1/mod.ts?s=WebhookOptions#prop_secretToken if you weren't aware.)

**Comment by chrisandrewcl** [comment]
@KnorpelSenf Installing with npm directly from the branch doesn't seem to work because there is a build step to generate the `./out` files needed on `node_modules`, so the installed package ends up unusable. I've tested locally by building with `deno task backport` and then installing it on my test project. I've double-checked that the installed module comes from my development version (e.g. by changing something in grammy and verifying it is reflected on my test project). Did I miss something in the process?

Regarding the `secretToken`: it should work, but it doesn't in my case because the handler is receiving the headers normalized to lowercase. I suspect this is by interference of another framework I am experimenting with ([SST](https://github.com/serverless-stack/sst), which adds a transparent wrapper around the lambda to facilitate development) since the documented behavior for aws lambdas is to handle headers case-sensitive, so they should have been received as is (i.e. `X-Telegram-Bot-Api-Secret-Token` instead of `x-telegram-bot-api-secret-token`). 

Let me know if you think more tests/changes are needed, and I will try to get to them in the coming days.

**Comment by chrisandrewcl** [comment]
Update regarding the `secretToken`: it seems I've misinterpreted the AWS docs. It doesn't seem to state anywhere that headers are delivered as is, just that they are "processed case-sensitive". To check it, I deployed a simple function that just print the received headers and during my test all headers were normalized lowercase. This behavior is not consistent with what I have previously encountered in other projects. I've also encountered similar confusion and multiple reports of conflicting and inconsistent behavior regarding how headers are passed thru lambda handlers on related repositories and other media.

In short, it seem we can't just pick one. I updated the PR to address this behavior.

Let me know what you think.

**Comment by KnorpelSenf** [comment]
> @KnorpelSenf Installing with npm directly from the branch doesn't seem to work because there is a build step to generate the `./out` files needed on `node_modules`, so the installed package ends up unusable. I've tested locally by building with `deno task backport` and then installing it on my test project. I've double-checked that the installed module comes from my development version (e.g. by changing something in grammy and verifying it is reflected on my test project). Did I miss something in the process?

That is very strange. If you install this package with npm from this repo, then it should run the prepare script after installation. This will run the backporting script. I'm surprised it doesn't work that way on your end … 

> ~Regarding the `secretToken`: it should work, but it doesn't in my case because the handler is receiving the headers normalized to lowercase. I suspect this is by interference of another framework I am experimenting with ([SST](https://github.com/serverless-stack/sst), which adds a transparent wrapper around the lambda to facilitate development) since the documented behavior for aws lambdas is to handle headers case-sensitive, so they should have been received as is (i.e. `X-Telegram-Bot-Api-Secret-Token` instead of `x-telegram-bot-api-secret-token`).~
> 
> Let me know if you think more tests/changes are needed, and I will try to get to them in the coming days.

So do you think we should add a new adapter for SST?

**Comment by chrisandrewcl** [comment]
> So do you think we should add a new adapter for SST?

No, not really. Since posting that comment I've done more tests and some research and posted my findings [here](https://github.com/grammyjs/grammY/pull/418#issuecomment-1567660003).

File: src/convenience/frameworks.ts [hunk 1]
```diff
@@ -178,13 +179,34 @@ const http: FrameworkAdapter = (req, res) => {
...
-    header: event.headers[SECRET_HEADER],
-    end: () => callback(null, { statusCode: 200 }),
-    respond: (json) => callback(null, { statusCode: 200, body: json }),
-    unauthorized: () => callback(null, { statusCode: 401 }),
-});
+const awsLambda: FrameworkAdapter = (event, _context, callback) => {
+    let resolveResponse: (response: unknown) => void;
+    let handlerReturn: Promise<unknown> | undefined;
+
+    if (callback) {
```
[the remark this answers is not in the document]
Comment (chrisandrewcl) [reply src/convenience/frameworks.ts:186]: AWS lambdas [support both ways](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html), so that is not uncommon for middleware, plugins, and extensions aimed at it.

My intention was to keep the legacy behavior while adding support for the modern format.

That being said, separating it as a new adapter is also an alternative. Maybe call it `aws-lambda-async`?

**Comment by KnorpelSenf** [comment]
> In short, it seem we can't just pick one. I updated the PR to address this behavior.
> 
> Let me know what you think.

Have you actually encountered a non-lowercase header for a grammY bot on AWS?

**Comment by chrisandrewcl** [comment]
> Have you actually encountered a non-lowercase header for a grammY bot on AWS?

During my tests, no. If you think it is ok to look only for the lowercase version, fine by me, I was just avoiding introducing breaking behavior.

File: src/convenience/frameworks.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (KnorpelSenf) [reply src/convenience/frameworks.ts:186]: Yes. For every call signature there's exactly one adapter.

For comparison, check the two cloudflare adapters. One is for module workers, and one is for legacy workers.

**Comment by KnorpelSenf** [comment]
It's good to be cautious about that. In this particular case, I'm fine with using the lowercase header name only. I see no way how it could change from request to request. If I'm wrong, it will be reliable to reproduce and simple to fix.

**Comment by chrisandrewcl** [comment]
> That is very strange. If you install this package with npm from this repo, then it should run the prepare script after installation. This will run the backporting script. I'm surprised it doesn't work that way on your end …

Ah, I see what happened. The project structure currently doesn't support being installed from git using `yarn`. I will switch to `npm` the next time I update the PR. Thanks!

File: src/convenience/frameworks.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (KnorpelSenf) [reply src/convenience/frameworks.ts:186]: Does the current one have a problem with secret tokens? You can fix it if that's the case. If there's no problem, there's no need to touch the code.

**Comment by KnorpelSenf** [comment]
Can we merge? Is there anything left to test or to fix?

**Comment by KnorpelSenf** [comment]
Ah nvm, we're waiting for https://github.com/grammyjs/grammY/pull/418#discussion_r1210229907

**Comment by chrisandrewcl** [comment]
> Can we merge? Is there anything left to test or to fix?

All done!

**Comment by KnorpelSenf** [comment]
@all-contributors add @chrisandrewcl for code
